core/ecschema-metadata/src/test/Metadata/CustomAttributeClass.test.ts

```diff
@@ -315,6 +315,87 @@ describe("CustomAttributeClass", () => {
 
     const newDom = createEmptyXmlDocument();
 
+    it("should properly serialize custom attributes having struct inheritance", async () => {
+      const schema = {
+        $schema: ECSchemaNamespaceUris.SCHEMAURL3_2_JSON,
+        name: "TestSchema",
+        version: "1.0.0",
+        alias: "ts",
+        customAttributes: [
+          {
+            className: "TestSchema.HasColors",
+            color: {
+              r: 40,
+              g: 128,
+              b: 68,
+              hexColor: "#288044",
+            },
+          },
+        ],
+        items: {
+          HexColor: {
+            schemaItemType: "StructClass",
+            properties: [
+            {
+              type: "PrimitiveProperty",
+              typeName: "string",
+              name: "hexColor",
+            },
+            ],
+          },
+          RgbColor: {
+            schemaItemType: "StructClass",
+            baseClass: "TestSchema.HexColor",
+            properties: [
+            {
+              type: "PrimitiveProperty",
+              typeName: "int",
+              name: "r",
+            },
+            {
+              type: "PrimitiveProperty",
+              typeName: "int",
+              name: "g",
+            },
+            {
+              type: "PrimitiveProperty",
+              typeName: "int",
+              name: "b",
+            },
+            ],
+          },
+          HasColors: {
+            schemaItemType: "CustomAttributeClass",
+            appliesTo: "Schema",
+            properties: [
+            {
+              name: "color",
+              type: "StructProperty",
+              typeName: "TestSchema.RgbColor",
+            },
+            ],
+          },
+        },
+      };
+
+      const ecschema = await Schema.fromJson(schema, new SchemaContext());
+      assert.isDefined(ecschema);
+
+      const serialized = await ecschema.toXml(newDom);
+      assert.isDefined(serialized);
+      const elements = serialized.lastChild;
```

Comment:
Even though this works, I don't like this code for asserting. It's very hard to read, especially since there is no XML snippet, understanding what the lastElement/firstElement refers to is not easy.
Imagine somebody getting a failure in this test because structProps.childNode does not have length 4. Troubleshooting this seems like a nightmare.

I suggest to load the xml back into a schema. Should be simple as it has no dependencies. Then you can use our API to get the custom attribute and check the values.

Reply:
Or serialize the schema to xml and compare it with an expected value. That's even easier to understand and easier to spot if something is missing.